### PR TITLE
[Sema] Fix a couple of pattern type-checking bugs

### LINKED ISSUE #63750: Duplicate bound variables not diagnosed in multi-statement closures
We incorrectly allow the following to compile:

```swift
let fn = {
  for (x,x) in [(0,0)] {}
  if case let (x,x) = (0,0) {}
  switch (0,0) {
  case let (x,x):
    ()
  }
}
```

we ought to diagnose duplicate bound vars, as we do outside of multi-statement closures.

### PR DESCRIPTION
**Pull request by hamishknight**
Avoid checking a switch for exhaustiveness in a multi-statement closure if the solution failed to apply, and ensure we always run redeclaration checking for bound pattern vars.

rdar://105781521
Resolves #63750

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Review by ahoppen — APPROVED** [review]
LGTM. Thanks

File: lib/Sema/CSSyntacticElement.cpp [hunk 1]
```diff
@@ -1738,9 +1738,10 @@ class SyntacticElementSolutionApplication
       }
     }
 
-    TypeChecker::checkSwitchExhaustiveness(
-        switchStmt, context.getAsDeclContext(), limitExhaustivityChecks);
-
+    if (!hadError) {
+      TypeChecker::checkSwitchExhaustiveness(
+          switchStmt, context.getAsDeclContext(), limitExhaustivityChecks);
```
Comment (CodaFi) [inline lib/Sema/CSSyntacticElement.cpp:1743]: I would just check exhaustiveness anyways but throw the limited mode switch so folks don’t have to build twice to see _something_.

Comment (hamishknight) [reply lib/Sema/CSSyntacticElement.cpp]: Makes sense, thanks!

**Comment by hamishknight** [comment]
@swift-ci please test

**Review by xedin — APPROVED** [review]

**Pull request merged**